text/0000-match-ergonmics.md

```diff
@@ -0,0 +1,531 @@
+method (if there is a `Deref` implementation) for any intermediate type found
+whilst dereferencing.
+
+We do not permit auto-referencing. Since `deref` is implemented for `&self`, to
+apply the auto-deref, `expr` must be a borrowed reference. E.g., if `expr` has
+type `Rc<Foo>` we will not implicitly dereference to `Foo`, but we will if the
+type is `&Rc<Foo>`. This means that we never implicitly borrow owned data, which
+should make this conversion easier to reason about. It also matches the
+behaviour of deref coercions (although it is weaker than the dot operator
+conversions).
+
+A less conservative alternative would be to allow auto-referencing. I believe
+there are no technical obstacles. That would be more flexible and match the
+behaviour of the dot operator. However, I believe constraining to borrowed data
+makes reasoning easier. We can always adopt auto-referencing later.
+
+The above implies that the patterns in each arm must have the same type. For
+example, the following match would not type check:
+
+```
+match x: &Option<_> {
+    Some(_) => {}
+    &None => {}
+}
+```
+
+
+### Type inference and backwards compatibility
+
+Currently, we might infer the type of the target expression from the type of the
+patterns. In order to be backwards compatible, we must be able to fall back to
+the type known from the patterns if type inference is stuck without that
+information. I think this will only require choosing `m = 0` if type inference
+can't make progress.
+
+
+## Match arm 'auto-ref'
+
+The programmer should no longer have to annotate variables in
+patterns with `ref` or `ref mut`; the compiler will infer that information from
+the arm body following the same rules as for captured variables in closures (see
+RFCs [114](https://github.com/rust-lang/rfcs/blob/master/text/0114-closures.md)
+and [231](https://github.com/rust-lang/rfcs/blob/master/text/0231-upvar-capture-inference.md)).
+
+For each variable declared in a pattern, the compiler will inspect how that
+variable is used. If the variable is required to be treated by value, then the
+value is moved or copied. If the value is mutated (or a mutable reference to the
+variable is taken), then the variable is mutably borrowed. Otherwise, the
+variable is immutably borrowed. Similarly to closures, the type of the variable
+is not affected, only its move/borrow semantics at runtime and in the borrow
+checker.
+
+In designing this feature we must be careful to ensure backwards compatibility
+where the user does write `ref`, `ref mut`, or `mut`.
+
+If there is a `ref` or `ref mut` annotation on a variable, then that annotation
+is obeyed and no further auto-referencing is performed. The compiler will not
+infer `mut` for a by-value variable even if it is mutated.
+
+Example with mutability:
+
+```
+let mut i: i32 = 0;
+match i {
+    j => j += 1,  // `j` is treated as `ref mut j` since `+=` takes `j` as `&mut i32`.
```

Comment:
> Is it your intent that users would have to write mut x when they want to move x

No, my belief is that 'inference' of move/borrow should be perfect (c.f., closures) and so `move` or forcing moving a single variable is not necessary. I agree that would be the wrong thing to do.

> If we want users to be able to force a by-value binding, ... It seems separate from inferring mut x to a ref.

You might be right (I admit, I haven't thought this through as thoroughly as the actual proposal. My worry is that with mutability, there is another category of use that doesn't exist without it. Consider

```
match x {
    mut y => { y.f = 0 }
}
```

Should this mutate `x` or not (or, IOW should `x` be moved)? There is no guide for the compiler whether `y` should be borrowed or not (both would compile) and the semantics are clearly different. If we do adopt the `mut` requirement and infer this as `ref mut` we break back compat, as far as I can see.